### PR TITLE
prov/efa: two patches related to fork support

### PR DESCRIPTION
**Pull request by wzamazon**
1.  replace ibv_fork_init() with aborting at fork()
2. do not unset RDMAV_HUGEPAGES_SAFE if user has it

### CODE REVIEW

File: prov/efa/src/efa_fabric.c [hunk 1]
```diff
@@ -907,25 +944,19 @@ int efa_fabric(struct fi_fabric_attr *attr, struct fid_fabric **fabric_fid,
...
 
 	/*
-	 * Enable rdma-core fork support and huge page support. We want call
-	 * this only when the EFA provider is selected. It is safe to call this
-	 * function again if multiple EFA fabrics are opened or if the fabric
-	 * is closed and opened again.
-	 *
-	 * TODO: allow users to disable this once the fork() to check ptrace
-	 * permissions is removed.
+	 * setting RDMAV_HUGEPAGES_SAFE along will not impact
```
Comment (rajachan) [inline prov/efa/src/efa_fabric.c]: Typo: along->alone?

File: prov/efa/src/efa_fabric.c [hunk 2]
```diff
@@ -907,25 +944,19 @@ int efa_fabric(struct fi_fabric_attr *attr, struct fid_fabric **fabric_fid,
...
-	ret = ibv_fork_init();
+	ret = pthread_atfork(efa_atfork_callback, NULL, NULL);
 	if (ret) {
-		EFA_WARN(FI_LOG_FABRIC, "Failed to initialize libibverbs "
-					"fork support. Please check your "
-					"application to ensure it is not "
-					"making verbs calls before "
-					"initializing EFA.\n");
+		EFA_WARN(FI_LOG_FABRIC,
+			 "Unable to register atfork callback");
```
Comment (rajachan) [inline prov/efa/src/efa_fabric.c:966]: Nit: Missing a line break.

File: prov/efa/src/efa_fabric.c [hunk 3]
```diff
@@ -899,6 +899,43 @@ static struct fi_ops_fabric efa_ops_fabric = {
...
+		"\n"
+		"However, setting this environment variable can result in\n"
+		"signficant performance impact to your application due to\n"
+		"increased cost of memory registration.\n"
+		"\n"
+		"You may want to check with your application vendor to see\n"
+		"if an application-level alternative (of not using fork)\n"
+		"exists.\n"
+		"\n"
+		"You job is aborted.\n");
```
Comment (rajachan) [inline prov/efa/src/efa_fabric.c:938]: Typo and nit: `You job is aborted` -> `Your job will now abort.`

File: prov/efa/src/efa_fabric.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/src/efa_fabric.c]: Thank you! fixed in new revision.

File: prov/efa/src/efa_fabric.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/efa_fabric.c]: Thank you! fixed in new revision.

File: prov/efa/src/efa_fabric.c (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/efa_fabric.c]: Thank you! fixed in new revision

File: prov/efa/src/efa_fabric.c (hunk 3 above)
Comment (wzamazon) [reply prov/efa/src/efa_fabric.c]: fixed the typo again.

File: prov/efa/src/efa_fabric.c [hunk 4]
```diff
@@ -907,25 +944,19 @@ int efa_fabric(struct fi_fabric_attr *attr, struct fid_fabric **fabric_fid,
...
 	/*
-	 * Enable rdma-core fork support and huge page support. We want call
-	 * this only when the EFA provider is selected. It is safe to call this
-	 * function again if multiple EFA fabrics are opened or if the fabric
-	 * is closed and opened again.
-	 *
-	 * TODO: allow users to disable this once the fork() to check ptrace
-	 * permissions is removed.
+	 * setting RDMAV_HUGEPAGES_SAFE alone will not impact
+	 * application peformance. It is only effective
```
Comment (bwbarrett) [inline prov/efa/src/efa_fabric.c]: "It is only effective" is ambiguous.  Maybe "rdma-core only checks this value when ..." or something similar?

File: prov/efa/src/efa_fabric.c [hunk 5]
```diff
@@ -899,6 +899,43 @@ static struct fi_ops_fabric efa_ops_fabric = {
...
+		"\n"
+		"However, setting this environment variable can result in\n"
+		"signficant performance impact to your application due to\n"
+		"increased cost of memory registration.\n"
+		"\n"
+		"You may want to check with your application vendor to see\n"
+		"if an application-level alternative (of not using fork)\n"
+		"exists.\n"
+		"\n"
+		"Your job will not abort.\n");
```
Comment (bwbarrett) [inline prov/efa/src/efa_fabric.c:938]: will now abort

File: prov/efa/src/efa_fabric.c (hunk 4 above)
Comment (wzamazon) [reply prov/efa/src/efa_fabric.c]: Thank you! fixed in new revision.

File: prov/efa/src/efa_fabric.c (hunk 5 above)
Comment (wzamazon) [reply prov/efa/src/efa_fabric.c]: Thank you! fixed in new revision.

**Comment by wzamazon** [comment]
bot:aws:retest

**Comment by wzamazon** [comment]
TravisCI error is something with the sockets provider

```
name:   fi_shared_ctx -e msg --no-rx-shared-ctx -p "sockets"
9823  result: Fail
9824  time:   2
9825  server_cmd: fi_shared_ctx -e msg --no-rx-shared-ctx -p "sockets"  -s 127.0.0.1
9826  server_stdout: |
9827    =================================================================
9828    ==88066==ERROR: AddressSanitizer: heap-use-after-free on address 0x6270000072e0 at pc 0x00010ae1448c bp 0x70000e473c30 sp 0x70000e473c28
9829    READ of size 4 at 0x6270000072e0 thread T4
9830        #0 0x10ae1448b in sock_conn_listener_thread (libfabric.1.dylib:x86_64+0xdb48b)
9831        #1 0x7fff5e1c7660 in _pthread_body (libsystem_pthread.dylib:x86_64+0x3660)
9832        #2 0x7fff5e1c750c in _pthread_start (libsystem_pthread.dylib:x86_64+0x350c)
9833        #3 0x7fff5e1c6bf8 in thread_start (libsystem_pthread.dylib:x86_64+0x2bf8)
9834    
9835    0x6270000072e0 is located 480 bytes inside of 12944-byte region [0x627000007100,0x62700000a390)
9836    freed by thread T0 here:
9837        #0 0x10afdca9d in wrap_free (libclang_rt.asan_osx_dynamic.dylib:x86_64+0x56a9d)
9838        #1 0x10adef2ff in sock_ep_close (libfabric.1.dylib:x86_64+0xb62ff)
9839        #2 0x10ace7bff in fi_close fabric.h:589
9840        #3 0x10ace738c in main shared_ctx.c:649
9841        #4 0x7fff5deaf014 in start (libdyld.dylib:x86_64+0x1014)
9842    
9843    previously allocated by thread T0 here:
9844        #0 0x10afdce27 in wrap_calloc (libclang_rt.asan_osx_dynamic.dylib:x86_64+0x56e27)
9845        #1 0x10adf4a34 in sock_alloc_endpoint (libfabric.1.dylib:x86_64+0xbba34)
9846        #2 0x10ade5c1b in sock_msg_ep (libfabric.1.dylib:x86_64+0xacc1b)
9847        #3 0x10aceae03 in fi_endpoint fi_endpoint.h:164
9848        #4 0x10ace8fb0 in server_connect shared_ctx.c:471
9849        #5 0x10ace7b2a in run shared_ctx.c:573
9850        #6 0x10ace729e in main shared_ctx.c:647
9851        #7 0x7fff5deaf014 in start (libdyld.dylib:x86_64+0x1014)
9852    
9853    Thread T4 created by T0 here:
9854        #0 0x10afd499d in wrap_pthread_create (libclang_rt.asan_osx_dynamic.dylib:x86_64+0x4e99d)
9855        #1 0x10ae139eb in sock_conn_start_listener_thread (libfabric.1.dylib:x86_64+0xda9eb)
9856        #2 0x10add5af2 in sock_domain (libfabric.1.dylib:x86_64+0x9caf2)
9857        #3 0x10aceb943 in fi_domain fi_domain.h:286
9858        #4 0x10ace8e0f in server_connect shared_ctx.c:460
9859        #5 0x10ace7b2a in run shared_ctx.c:573
9860        #6 0x10ace729e in main shared_ctx.c:647
9861        #7 0x7fff5deaf014 in start (libdyld.dylib:x86_64+0x1014)
9862    
9863    SUMMARY: AddressSanitizer: heap-use-after-free (libfabric.1.dylib:x86_64+0xdb48b) in sock_conn_listener_thread
9864    Shadow bytes around the buggy address:
9865      0x1c4e00000e00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
9866      0x1c4e00000e10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
9867      0x1c4e00000e20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
9868      0x1c4e00000e30: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
9869      0x1c4e00000e40: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
9870    =>0x1c4e00000e50: fd fd fd fd fd fd fd fd fd fd fd fd[fd]fd fd fd
9871      0x1c4e00000e60: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
9872      0x1c4e00000e70: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
9873      0x1c4e00000e80: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
9874      0x1c4e00000e90: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
9875      0x1c4e00000ea0: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
9876    Shadow byte legend (one shadow byte represents 8 application bytes):
9877      Addressable:           00
9878      Partially addressable: 01 02 03 04 05 06 07 
9879      Heap left redzone:       fa
9880      Freed heap region:       fd
9881      Stack left redzone:      f1
9882      Stack mid redzone:       f2
9883      Stack right redzone:     f3
9884      Stack after return:      f5
9885      Stack use after scope:   f8
9886      Global redzone:          f9
9887      Global init order:       f6
9888      Poisoned by user:        f7
9889      Container overflow:      fc
9890      Array cookie:            ac
9891      Intra object redzone:    bb
9892      ASan internal:           fe
9893      Left alloca redzone:     ca
9894      Right alloca redzone:    cb
9895    ==88066==ABORTING
9896    Posting recv for endpoint #0
9897    Posting recv for endpoint #1
9898    Posting recv for endpoint #2
9899    Posting recv for endpoint #3
9900    Posting send #0 to shared tx ctx
9901    Posting send #1 to shared tx ctx
9902    Posting send #2 to shared tx ctx
9903    Posting send #3 to shared tx ctx
9904    ./scripts/runfabtests.sh: line 434: 88065 Abort trap: 6           timeout 120 fi_shared_ctx -e msg --no-rx-shared-ctx -p "sockets" -s 127.0.0.1
9905  client_cmd: fi_shared_ctx -e msg --no-rx-shared-ctx -p "sockets"  -s 127.0.0.1 127.0.0.1
9906  client_stdout: |
9907    Posting recv for endpoint #0
9908    Posting recv for endpoint #1
9909    Posting recv for endpoint #2
9910    Posting recv for endpoint #3
9911    Posting send #0 to shared tx ctx
9912    Posting send #1 to shared tx ctx
9913    Posting send #2 to shared tx ctx
9914    Posting send #3 to shared tx ctx
```
should not be related

**Pull request merged**